text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+- **Assets**: any files that must be available in the final built application directory such that they have public URLs (typical examples are images and fonts).
+- **Build Hooks**: code that runs within Node at application build time. The v1 equivalent is an addon's `index.js` file.
+
+## Own Javascript
+
+The public `main` (as defined in `package.json`) of a v2 package points to its **Own Javascript**. The code is formatted as ES modules using ES latest features, meaning: stage 4 features only, unless otherwise explicitly mentioned elsewhere in this spec. (Addon authors can still use whatever custom syntax they want, but those babel plugins must run before publication to NPM.)
+
+Templates are in hbs format. No custom AST transforms are supported. (Addon authors can still use whatever custom AST transforms they want, but those transforms must have already been applied before publication to NPM.)
+
+Unlike v1 addons, there is no `/app` or `/addon` directory that is magically removed from the runtime paths to the modules. All resolution follows the prevailing Node rules.
+
+The benefit of this design is that it makes our packages understandable by a broader set of tooling. Editors and build tools can follow `import` statements across packages and end up in the right place.
+
+In v1 packages, `main` usually points to a build-time configuration file. That file is moving and will be described in the **Build Hooks** section below.
+
+### Own Javascript: Imports
+
+Modules in **Own Javascript** are allowed to use ECMA static `import` to resolve any **allowed dependency**, causing it to be included in the build whenever the importing module is included. This replaces `app.import`. Resolution follows prevailing Node rules. (This usually means the node_modules algorithm, but it could also mean Yarn PnP. The difference shouldn't matter if you are correctly declaring all your **allowed dependencies**.)
+
+Notice that a package’s **allowed dependencies** do not include the package itself. This is consistent with how Node resolution works. To import files from within your own package you must use relative paths. This is different from how run-time AMD module resolution has historically worked in Ember Apps. (`@embroider/compat` implements automatic adjustment for this case when compiling from v1 to v2).
+
+Modules in **Own Javascript** are allowed to use dynamic `import()`, and the specifiers have the same meanings as in static import. However, we impose a restriction on what syntax is supported inside `import()`. The only supported syntax inside `import()` is:
+
+- a string-literal
+- or a template string literal
+  - that begins with a static part
+  - where the static part is required to contain at least one `/`
+  - or at least two `/` when the path starts with `@`.
```

Comment:
This should mention dynamic imports from external URLs explicitly. They are different from other imports as they aren't linked at build-time. It's currently unclear to me if they are supported and what supported means in this context.

There are three different types of dynamic imports from external URLs:

1. From a static URL
   ```js
   import('https://examples.com/foo.js');
   ```
2. From a URL template
   ```js
   import(`https://examples.com/${file}`);
   import(`https://${url}`);
   import(`//${url}`);
   ```
3. From a variable
   ```js
   let url = 'https://examples.com/foo.js');
   import(url);
   ```

Only for first and second example it's known at build-time that the target of the dynamic import is an external URL. The compiler does not know anything about the third example.

All three examples are valid JavaScript. They are [supported natively in most browsers](https://caniuse.com/#feat=es6-module-dynamic-import). Only IE11 and Edge versions that are not based on Chrome yet do not support dynamic imports.

They work in current Ember applications unless the app uses ember-auto-import. Ember-auto-import throws a build-time error for second and third example that the import target must be a string literal. If the app does not use ember-auto-import the dynamic import will not be transpiled and included in the build untouched.

The limitation of ember-auto-import was [raised in this issue](https://github.com/ef4/ember-auto-import/issues/276). Four possible solutions were discussed in the issue:

1. Support template literals if they start with `https://` or `//`.
   ```js
   import(`https://${url}`);
   import(`//${url}`);
   ```
2. Skip build-time transform with magic comments.
   ```js
   /* ember-auto-import-disable */
   import(url);
   /* ember-auto-import-enable */
   ```
3. Reserve `import()` for runtime imports and add a dedicated build-time-`import()` macro.
4. Add a lint rule forbidding dynamic imports. Silently skip dynamic imports at build-time and leave them untouched instead of throwing.

The first option would break some valid JavaScript. It may even break existing Ember applications. But it would still allow all use cases cause the [list of possible protocols is known upfront](https://github.com/ef4/ember-auto-import/issues/276#issuecomment-611597946).

The recommended solution for webpack [seems to be](https://github.com/webpack/webpack/issues/8341) magic comments: `import(/* webpackIgnore: true */ "/js/hello.js")`. But [@ef4 explained](https://github.com/ef4/ember-auto-import/issues/276#issuecomment-613087368) that it's the most complex option from implementation perspective.

The dedicated macro for build-time dynamic imports would be the simplest option from implementation perspective. But it would be different from the rest of the JS ecosystem and therefore increases the teaching effort dramatically as pointed out by @ef4 in the same comment.

The last option would be a change to the overall strategy. A dynamic import which target is not known at build-time would not trigger a build-time error but would be silently skipped. Instead the developer would be warned by a lint rule. It may cause strange and hard to debug issues if the developer ignores such a lint error. But valid JavaScript would just work. To be honest this is my preferred option.

A potential use case for dynamic imports from external URLs is expanding the components available to an application at run-time. E.g. there might be a registry of web components to which the user can publish new components at any time. The application does not know which web components exists at build-time. But it can resolve all of them using dynamic imports from an external URL at run-time.

Reply:
I have now updated the document to reflect that we would allow URLs as long as we can tell from their static prefix that they're URLs.